File: statistics/stats.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Sigmath
{
    public class Statistics
    {
        public static double Mean(IEnumerable<double> values)
        {
            if (values == null || !values.Any())
            {
                throw new ArgumentException("Cannot calculate mean of empty sequence");
            }

            var count = 0;
            var sum = 0.0;

            foreach (var value in values)
            {
                count++;
                sum += value;
            }

            return sum / count;
        }

        public static double Median(IEnumerable<double> values)
        {
            if (values == null || !values.Any())
            {
                throw new ArgumentException("Cannot calculate median of empty sequence");
            }

            var sortedValues = values.OrderBy(v => v).ToArray();
            var middleIndex = sortedValues.Length / 2;

            if (sortedValues.Length % 2 == 0)
            {
                return (sortedValues[middleIndex - 1] + sortedValues[middleIndex]) / 2.0;
            }
            else
            {
                return sortedValues[middleIndex];
            }
        }

        public static double Mode(IEnumerable<double> values)
        {
            if (values == null || !values.Any())
            {
                throw new ArgumentException("Cannot calculate mode of empty sequence");
            }

            var valueCounts = new Dictionary<double, int>();

            foreach (var value in values)
            {
                if (!valueCounts.ContainsKey(value))
                {
                    valueCounts.Add(value, 0);
                }

                valueCounts[value]++;
            }

            var maxCount = valueCounts.Values.Max();
            var modes = valueCounts.Where(pair => pair.Value == maxCount).Select(pair => pair.Key);

            if (modes.Count() == 1)
            {
                return modes.First();
            }
            else
            {
                throw new InvalidOperationException("Sequence has multiple modes");
            }
        }

        public static double Variance(IEnumerable<double> values)
        {
            if (values == null || !values.Any())
            {
                throw new ArgumentException("Cannot calculate variance of empty sequence");
            }

            var mean = Mean(values);
            var count = 0;
            var sumOfSquaredDifferences = 0.0;

            foreach (var value in values)
            {
                count++;
                var difference = value - mean;
                sumOfSquaredDifferences += difference * difference;
            }

            return sumOfSquaredDifferences / count;
        }

        public static double StandardDeviation(IEnumerable<double> values)
        {
            return Basic.SquareRoot(Variance(values));
        }

        public static double Correlation(IEnumerable<double> valuesX, IEnumerable<double> valuesY)
        {
            if (valuesX == null || valuesY == null || !valuesX.Any() || !valuesY.Any())
            {
                throw new ArgumentException("Cannot calculate correlation coefficient of empty sequences");
            }

            var count = 0;
            var sumX = 0.0;
            var sumY = 0.0;
            var sumXY = 0.0;
            var sumXSquared = 0.0;
            var sumYSquared = 0.0;

            using (var enumX = valuesX.GetEnumerator())
            using (var enumY = valuesY.GetEnumerator())
            {
                while (enumX.MoveNext() && enumY.MoveNext())
                {
                    var valueX = enumX.Current;
                    var valueY = enumY.Current;
                    count++;
                    sumX += valueX;
                    sumY += valueY;
                    sumXY += valueX * valueY;
                    sumXSquared += valueX * valueX;
                    sumYSquared += valueY * valueY;
                }
            }

            var numerator = count * sumXY - sumX * sumY;
            var denominator = Basic.SquareRoot((count * sumXSquared - sumX * sumX) * (count * sumYSquared - sumY * sumY));

            return numerator / denominator;
        }

    }
}

File: Sigmath.ConvertedToVBNet/statistics/stats.vb
Imports System.Collections.Generic
Imports System.Linq

Public Class Statistics
	Public Shared Function Mean(values As IEnumerable(Of Double)) As Double
		If values Is Nothing OrElse Not values.Any() Then
			Throw New ArgumentException("Cannot calculate mean of empty sequence")
		End If

		Dim count = 0
		Dim sum = 0.0

		For Each value As var In values
			count += 1
			sum += value
		Next

		Return sum / count
	End Function

	Public Shared Function Median(values As IEnumerable(Of Double)) As Double
		If values Is Nothing OrElse Not values.Any() Then
			Throw New ArgumentException("Cannot calculate median of empty sequence")
		End If

		Dim sortedValues = values.OrderBy(Function(v) v).ToArray()
		Dim middleIndex = sortedValues.Length \ 2

		If sortedValues.Length Mod 2 = 0 Then
			Return (sortedValues(middleIndex - 1) + sortedValues(middleIndex)) / 2.0
		Else
			Return sortedValues(middleIndex)
		End If
	End Function

	Public Shared Function Mode(values As IEnumerable(Of Double)) As Double
		If values Is Nothing OrElse Not values.Any() Then
			Throw New ArgumentException("Cannot calculate mode of empty sequence")
		End If

		Dim valueCounts = New Dictionary(Of Double, Integer)()

		For Each value As var In values
			If Not valueCounts.ContainsKey(value) Then
				valueCounts.Add(value, 0)
			End If

			valueCounts(value) += 1
		Next

		Dim maxCount = valueCounts.Values.Max()
		Dim modes = valueCounts.Where(Function(pair) pair.Value = maxCount).[Select](Function(pair) pair.Key)

		If modes.Count() = 1 Then
			Return modes.First()
		Else
			Throw New InvalidOperationException("Sequence has multiple modes")
		End If
	End Function

	Public Shared Function Variance(values As IEnumerable(Of Double)) As Double
		If values Is Nothing OrElse Not values.Any() Then
			Throw New ArgumentException("Cannot calculate variance of empty sequence")
		End If

		Dim mean__1 = Mean(values)
		Dim count = 0
		Dim sumOfSquaredDifferences = 0.0

		For Each value As var In values
			count += 1
			Dim difference = value - mean__1
			sumOfSquaredDifferences += difference * difference
		Next

		Return sumOfSquaredDifferences / count
	End Function

	Public Shared Function StandardDeviation(values As IEnumerable(Of Double)) As Double
		Return Basic.SquareRoot(Variance(values))
	End Function

	Public Shared Function Correlation(valuesX As IEnumerable(Of Double), valuesY As IEnumerable(Of Double)) As Double
		If valuesX Is Nothing OrElse valuesY Is Nothing OrElse Not valuesX.Any() OrElse Not valuesY.Any() Then
			Throw New ArgumentException("Cannot calculate correlation coefficient of empty sequences")
		End If

		Dim count = 0
		Dim sumX = 0.0
		Dim sumY = 0.0
		Dim sumXY = 0.0
		Dim sumXSquared = 0.0
		Dim sumYSquared = 0.0

		Using enumX = valuesX.GetEnumerator()
			Using enumY = valuesY.GetEnumerator()
				While enumX.MoveNext() AndAlso enumY.MoveNext()
					Dim valueX = enumX.Current
					Dim valueY = enumY.Current
					count += 1
					sumX += valueX
					sumY += valueY
					sumXY += valueX * valueY
					sumXSquared += valueX * valueX
					sumYSquared += valueY * valueY
				End While
			End Using
		End Using

		Dim numerator = count * sumXY - sumX * sumY
		Dim denominator = Basic.SquareRoot((count * sumXSquared - sumX * sumX) * (count * sumYSquared - sumY * sumY))

		Return numerator / denominator
	End Function

End Class
